### PR TITLE
VDYP-306  Veteran Layer Processing for VRIStart

### PR DESCRIPTION
**Pull request by smithkm**

### CODE REVIEW

File: vdyp-lib/vdyp-fip/src/main/java/ca/bc/gov/nrs/vdyp/fip/FipStart.java.orig
Comment (smithkm) [inline vdyp-lib/vdyp-fip/src/main/java/ca/bc/gov/nrs/vdyp/fip/FipStart.java.orig:1]: I accidentally committed originals from a merge in a previous PR.  Here I'm removing them.

File: vdyp-lib/vdyp-fip/src/test/java/ca/bc/gov/nrs/vdyp/fip/FipStartTest.java.orig
Comment (smithkm) [inline vdyp-lib/vdyp-fip/src/test/java/ca/bc/gov/nrs/vdyp/fip/FipStartTest.java.orig:1]: I accidentally committed originals from a merge in a previous PR.  Here I'm removing them.

File: vdyp-lib/vdyp-common/src/main/java/ca/bc/gov/nrs/vdyp/application/VdypStartApplication.java [hunk 1]
```diff
@@ -1400,4 +1400,131 @@ protected void scaleAllSummableUtilization(VdypUtilizationHolder holder, float f
...
+					}
+
+					// Set component 0 to equal component 4.
+					utilVector.setAll(utilVector.getLarge());
+
+					accessors.getWriteMethod().invoke(vdypSpecies, utilVector);
+				}
+			}
+
+			computeLayerUtilizationComponentsFromSpecies(vdypLayer);
```
Comment (mjunkin) [inline vdyp-lib/vdyp-common/src/main/java/ca/bc/gov/nrs/vdyp/application/VdypStartApplication.java:1523]: Not sure where PCTLV (YUCV lines 171-173) is being calculated (assuming it needs to be...)

File: vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java [hunk 2]
```diff
@@ -631,11 +649,143 @@ float quadMeanDiameterSpeciesAdjust(float x, float initialDq, float min, float m
 		return FloatMath.clamp(7.5f + (initialDq - 7.5f) * FloatMath.exp(x), min, max);
 	}
 
-	private void processVeteranLayer(VriPolygon polygon, VdypLayer.Builder lBuilder) throws StandProcessingException {
+	private float processVeteranLayer(VriPolygon polygon, VdypLayer.Builder lBuilder) throws StandProcessingException {
 		@SuppressWarnings("unused")
```
Comment (mjunkin) [inline vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java:653]: veteranLayer is being used now...

File: vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java [hunk 3]
```diff
@@ -631,11 +649,143 @@ float quadMeanDiameterSpeciesAdjust(float x, float initialDq, float min, float m
...
 		var veteranLayer = polygon.getLayers().get(LayerType.VETERAN);
+		lBuilder.layerType(LayerType.VETERAN);
+
+		lBuilder.adapt(veteranLayer);
+		
+		var bec = polygon.getBiogeoclimaticZone();
+
+		var lowDq = false;
```
Comment (mjunkin) [inline vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java:661]: Unused

File: vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java [hunk 4]
```diff
@@ -631,11 +649,143 @@ float quadMeanDiameterSpeciesAdjust(float x, float initialDq, float min, float m
...
+
+		var primarySite = veteranLayer.getPrimarySite();
+		var primarySpecies = veteranLayer.getPrimarySpeciesRecord();
+
+		float ageTotal = primarySite.flatMap(VriSite::getAgeTotal).map(at -> at + veteranLayer.getAgeIncrease())
+				.orElse(0f); // AGETOTLV
+		float yearsToBreastHeight = primarySite.flatMap(VriSite::getYearsToBreastHeight).orElse(0f); // YTBHLV
+		float breastHeightAge = ageTotal - yearsToBreastHeight; // AGEBHLV
```
Comment (mjunkin) [inline vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java:669]: Unused

File: vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java [hunk 5]
```diff
@@ -631,11 +649,143 @@ float quadMeanDiameterSpeciesAdjust(float x, float initialDq, float min, float m
...
+
+		lBuilder.buildChildren();
+		var specList = lBuilder.getSpecies();
+
+		// Really crude reconciliation to the specified TPH (if any)
+		// Could improve with code similar to ROOTV01
+		if (treesPerHectare > 0) {
+			float tphSum = 0;
+			for (var spec : specList) {
+				tphSum += spec.getTreesPerHectareByUtilization().getAll();
```
Comment (mjunkin) [inline vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java:722]: The FORTRAN code reads UC Large, not UC All. But of course they may be known to have the same value (although not to me...)

File: vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java [hunk 6]
```diff
@@ -631,11 +649,143 @@ float quadMeanDiameterSpeciesAdjust(float x, float initialDq, float min, float m
...
+			baSum += spec.getBaseAreaByUtilization().getAll();
+		}
+
+		if (polygon.getMode().filter(mode -> mode == PolygonMode.BATC).isPresent()) {
+			// for mode 4 (with CC) input TPH was NOT available.
+			inputTreesPerHectare = tphSum;
+			treesPerHectare = tphSum;
+		}
+
+		return inputTreesPerHectare;
```
Comment (mjunkin) [inline vdyp-lib/vdyp-vri/src/main/java/ca/bc/gov/nrs/vdyp/vri/VriStart.java:755]: There's this check (after the call to YUCV) in the original code - not sure where it is in your code:

      if ( abs (TPHinput- TPH(0,4))/TPHinput .gt. 0.0005)
                    LOW_DQ = .true.
      if (LOW_DQ) IER= -1

**Review by mjunkin — COMMENTED** [review]
There's a few questions - please have a look and let me know

File: vdyp-lib/vdyp-common/src/main/java/ca/bc/gov/nrs/vdyp/application/VdypStartApplication.java (hunk 1 above)
Comment (smithkm) [reply vdyp-lib/vdyp-common/src/main/java/ca/bc/gov/nrs/vdyp/application/VdypStartApplication.java:1530]: https://github.com/bcgov/NR-VDYP/pull/57/commits/0c0a71e7e0c5a5835a30623dca455e1159744238

**Review by mjunkin — APPROVED** [review]

**Pull request merged**